microcosm_caching/decorators.py

```diff
@@ -118,3 +126,73 @@ def cache(*args, **kwargs) -> Schema:
 
         return cache
     return decorator
+
+
+def invalidates(
+    component,
+    invalidations,
+    cache_prefix,
+    lock_ttl=DEFAULT_LOCK_TTL
+):
+    """
+    Invalidates a set of prescribed keys, based on a combination of:
+        * specified arguments
+        * schema
+
+    Note: this does require that the input args to the given function correspond to the
+    invalidation args given. As such, using it conjunction with certain functions (such as
+    a plain delete) may not provide sufficient context for invalidation, and caching should
+    not be used in such scenarios.
+
+    """
+    graph = component.graph
+    metrics = get_metrics(graph)
+    resource_cache: CacheBase = graph.resource_cache
+
+    def delete_from_cache(values) -> None:
+        """
+        "Delete" from cache by locking writes to a key for a designated
+        amount of time.
+
+        In conjunction with the cached() decorator, this allows the "get and set"
+        flow to behave without special cases
+
+        """
+        start_time = perf_counter()
+
+        resource_cache.set_many(values, ttl=lock_ttl)
+
+        elapsed_ms = (perf_counter() - start_time) * 1000
+
+        if metrics:
+            tags = [
+                "action:set_many",
+            ]
+
+            metrics.timing("cache_timing", elapsed_ms, tags=tags)
+            metrics.increment("cache_set_many", tags=tags)
+
+    def decorator(func):
+        @wraps(func)
+        def cache(*args, **kwargs) -> Schema:
+            if not resource_cache:
+                return func(*args, **kwargs)
+
+            values: Dict[str, None] = {}
```

Comment:
not sure why you are batching here. should we just invalidate on line 188?

Reply:
You can imagine a world where one resource can have several dependent resources, and you would want to invalidate them all at once.

Reply:
yes, you are more likely to batch invalidate. the cache, you want to get it in there as fast as possible.. waiting for a batch to form increase the window for race conditions.